### ISSUE TITLE
i++

### ISSUE THREAD

**Issue by hakunin** [mirah/mirah#41]
I often find myself in need of i++, for example:

```
i = 0
rows.each { |row|
  row_class = (i % 2 == 0) ? "even" : "odd"
  html += "..."
  i += 1
}
```

Writing i += 1 sucks. Also i++ seems to be more expressive because I don't mean to _add one_, I man to _increment_.

**Comment by felixvf** [comment]
I’d like to close this as "WONTFIX".

Ruby does not have "i++", Crystal does not have "i++" and

```mirah
f(i++,a[i++],i++)[i++] = b[i++] + c[i++]
```

is not well defined.

In addition, unlike some languages close to machine code (C), there is no performance optimization to say "i++". Instead, "i++" will always have to translate to the bytecode of "i+=1".

In addition, there is no clear meaning of "++"-style postfix operators. Should `i++` mean the same as `i = i.+(1)`? If so, should `i--` mean the same as `i = i.-(1)`? If so, should `i**` mean the same as `i = i.*(1)`? If so, should `i//` mean the same as `i = i./(1)`? If so, should `i&&` mean the same as `i = i.&(1)`?

Apparently, it is not useful to extend this concept to operators beyond "++" or "--". But if the concept is not generic enough, is its ungenericness, its peculiarity, its oddity warranted by its usefulness? I doubt that. In C and C++, "++" works on integers and pointers. In Mirah, we do not have pointer arithmetic. And we do not need explicit incrementing as often as in C, as we have `#each`, `#map` and other ways of iterating.

So I’m up for closing this.

However, if there is a strong need, I’m open for patches which make postfix "++" and postfix "--" an operator, so users can define their own meaning of "++" by defining a macro `int#++` if the really want to.

**Comment by baroquebobcat** [comment]
I agree. I've been reading some C recently and I'd say postfix ++ has some
allowed usages that end up being really confusing.

On May 12, 2017 7:05 PM, "Felix von Ferey" <notifications@github.com> wrote:

I’d like to close this as "WONTFIX".

Ruby does not have "i++", Crystal does not have "i++" and

f(i++,a[i++],i++)[i++] = b[i++] + c[i++]

is not well defined.

In addition, unlike some languages close to machine code (C), there is no
performance optimization to say "i++". Instead, "i++" will always have to
translate to the bytecode of "i+=1".

In addition, there is no clear meaning of "++"-style postfix operators.
Should i++ mean the same as i = i.+(1)? If so, should i-- mean the same as i
= i.-(1)? If so, should i** mean the same as i = i.*(1)? If so, should i//
mean the same as i = i./(1)? If so, should i&& mean the same as i = i.&(1)?

Apparently, it is not useful to extend this concept to operators beyond
"++" or "--". But if the concept is not generic enough, is its
ungenericness, its peculiarity, its oddity warranted by its usefulness? I
doubt that. In C and C++, "++" works on integers and pointers. In Mirah, we
do not have pointer arithmetic. And we do not need explicit incrementing as
often as in C, as we have #each, #map and other ways of iterating.

So I’m up for closing this.

However, if there is a strong need, I’m open for patches which make postfix
"++" and postfix "--" an operator, so users can define their own meaning of
"++" by defining a macro int#++ if the really want to.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/mirah/mirah/issues/41#issuecomment-301215647>, or mute
the thread
